docs/src/test-auth-js.md

```diff
@@ -0,0 +1,311 @@
+import { test } from '@playwright/test';
+
+test('test', async ({ page }) => {
+  // page is signed in.
+});
+```
+
+```js js-flavor=js
+const { test } = require('@playwright/test');
+
+test('test', async ({ page }) => {
+  // page is signed in.
+});
+```
+
+:::note
+If you can log in once and commit the `storageState.json` into the repository, you won't need the global
+setup at all, just specify the `storageState.json` in Playwright Config as above and it'll be picked up.
+:::
+
+### Multiple signed in roles
+
+Sometimes you have more than one signed-in user in your end to end tests. You can achieve that via logging in for these users multiple times in globalSetup and saving that state into different files.
+
+```js js-flavor=js
+// global-setup.js
+const { chromium } = require('@playwright/test');
+
+module.exports = async config => {
+  const browser = await chromium.launch();
+  const page = await browser.newPage();
+  await page.goto('https://github.com/login');
+  await page.fill('input[name="user"]', 'user');
+  await page.fill('input[name="password"]', 'password');
+  await page.click('text=Sign in');
+  await page.context().storageState({ path: 'storageState.json' });
+  await browser.close();
+};
+```
+
+```js js-flavor=ts
+// global-setup.ts
+import { chromium, FullConfig } from '@playwright/test';
+
+async function globalSetup(config: FullConfig) {
+  const browser = await chromium.launch();
+  const adminPage = await browser.newPage();
+  // ... log in
+  await adminPage.context().storageState({ path: 'adminStorageState.json' });
+
+  const userPage = await browser.newPage();
+  // ... log in
+  await userPage.context().storageState({ path: 'userStorageState.json' });
+  await browser.close();
+}
+
+export default globalSetup;
+```
+
+After that you can specify the user to use for each test file or each test group:
+
+```js js-flavor=ts
+import { test } from '@playwright/test';
+
+test.use({ storageState: 'adminStorageState.json' });
+
+test('admin test', async ({ page }) => {
+  // page is signed in as admin.
+});
+
+test.describe(() => {
+  test.use({ storageState: 'userStorageState.json' });
+
+  test('user test', async ({ page }) => {
+    // page is signed in as a user.
+  });
+});
+```
+
+```js js-flavor=js
+const { test } = require('@playwright/test');
+
+test.use({ storageState: 'adminStorageState.json' });
+
+test('admin test', async ({ page }) => {
+  // page is signed in as amin.
+});
+
+test.describe(() => {
+  test.use({ storageState: 'userStorageState.json' });
+
+  test('user test', async ({ page }) => {
+    // page is signed in as a user.
+  });
+});
+```
+
+### Reuse the signed in page in multiple tests
+
+Although discouraged, sometimes it is necessary to sacrifice the isolation and run a number of tests
+in the same page. In that case, you can log into that page once in `beforeAll` and then use that same
+page in all the tests. Note that you need to run these tests serially using `test.describe.serial` in
```

Comment:
Let's link to the serial/page reuse doc?

Reply:
Linking to retries feels artificial here, we should re-create core concepts and link there.